### DISCUSSION TITLE
Issue with Setting fcmOptions in Firebase PHP SDK for Cloud Messaging

### DISCUSSION THREAD

**Discussion by erdman1** [kreait/firebase-php#859] [Q&A] [answered]
### Problem Description
I'm facing a problem with the Firebase PHP SDK where `fcmOptions` in a `CloudMessage` object are not being set correctly. Despite setting `fcm_options` within `WebPushConfig`, the `fcmOptions` attribute of `CloudMessage` remains null, resulting in an `Invalid JSON payload received` error when attempting to send messages.

### Implementation Details
- `fcm_options` are set within `WebPushConfig::fromArray()`.
- `CloudMessage` is constructed with `withTarget()`, `withWebPushConfig()`, and `withFcmOptions()`.
- `FcmOptions` are created using `FcmOptions::fromArray()`.

### Expected Behavior
`fcmOptions` in `CloudMessage` should be populated and correctly recognized by Firebase.

### Environment
- PHP Version: 8.1.2
- Firebase PHP SDK Version: [Your SDK version]
- kreait/firebase-php  version:  5.19.0  

### Full Code Causing the Issue
```php
        $messaging = $this->firebaseService->getMessaging();
        $fcmOptions = FcmOptions::fromArray([
            'link' => $customNotificationData['post_url']
        ]);
        
        $message = CloudMessage::withTarget('token', $token)
        ->withNotification(['title' => 'Title', 'body' => 'Body'])
        ->withWebPushConfig(WebPushConfig::fromArray([
            'notification' => [
                'title' => 'Title',
                'body' => 'Body',
                'icon' => 'icon.png'
            ],
            'fcm_options' => [
                'link' => 'https://your-link.com'
            ]
        ]))
        ->withFcmOptions($fcmOptions);
        ;
      
    
    $messaging->send($message);

```

### Error Encountered
`Exception has occurred.
Fatal error: Uncaught Kreait\Firebase\Exception\Messaging\InvalidMessage: Invalid JSON payload received. Unknown name "link" at 'message.fcm_options': Cannot find field.`

I would appreciate any insights or solutions to resolve this issue.

**Comment by jeromegamez** [comment]
There are two different kinds of FCM Options: the one at the root level of a message only supports the `analytics_label` field, the WebPush Config one supports the `link` field, which you're already setting correctly. 

The SDK doesn't actively validate values, but when you use a static code analyser like PHPStan, it would have notified about the wrong field.

https://github.com/kreait/firebase-php/blob/7.x/src/Firebase/Messaging/FcmOptions.php d

I should add the static analysis information to the documentation, if haven't already 😅.

If you don't use an analytics label, you can Safely remove the `FcmOptions` object and the `withFcmOptions()` method call.

Let me know if this helps!

**Comment by erdman1** [comment]
Thanks, but even when I have it like the below, the link is notification is not clickable...

```php
        $messaging = $this->firebaseService->getMessaging();        
        $message = CloudMessage::withTarget('token', $token)
        ->withNotification(['title' => 'Title', 'body' => 'Body'])
        ->withWebPushConfig(WebPushConfig::fromArray([
            'notification' => [
                'title' => 'Title',
                'body' => 'Body',
                'icon' => 'icon.png'
            ],
            'fcm_options' => [
                'link' => 'https://your-link.com'
            ]
        ]))
        ;
      
    
    $messaging->send($message);
```



![Code_tIKP3KFlHw](https://github.com/kreait/firebase-php/assets/150612804/4becd468-d7dc-4973-bed2-1dc2e9a78e03)

**Comment by erdman1** [comment, marked as answer]
Okay, so after further review I was finally able to get to the bottom of this.

The php code now looks like this:

```php
        $messaging = $this->firebaseService->getMessaging();        
        $message = CloudMessage::withTarget('token', $token)
        ->withNotification(['title' => 'Title', 'body' => 'Body'])
        ->withWebPushConfig(WebPushConfig::fromArray([
            'notification' => [
                'title' => 'Title',
                'body' => 'Body',
                'icon' => 'icon.png'
            ],
            'fcm_options' => [
                'link' => 'http://localhost/test'
            ]
        ]));    
      
    
    $messaging->send($message);

```

my main "issue" was with the sw js not set correctly, now it looks like the below and it works great!

```js
self.addEventListener('notificationclick', function(event) {
    // Access the click_action URL from the notification data
    const clickAction = event.notification.data.FCM_MSG.notification.click_action;

    console.log('Notification click_action URL:', clickAction);

    // Close the notification
    event.notification.close();

    // Open the click_action URL in a new window
    event.waitUntil(
        clients.openWindow(clickAction)
    );
});
```

obviously still needs refinement, but this works
